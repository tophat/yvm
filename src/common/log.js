const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

function log(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

module.exports = log
