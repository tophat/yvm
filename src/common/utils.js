const os = require('os')
const path = require('path')
const request = require('request')

const log = require('../common/log')

const yvmPath = process.env.YVM_DIR || path.resolve(os.homedir(), '.yvm')
const versionRootPath = rootPath => path.resolve(rootPath, 'versions')

const getExtractionPath = (version, rootPath) =>
    path.resolve(rootPath, 'versions', `v${version}`)

const stripVersionPrefix = tagName =>
    tagName[0] === 'v' ? tagName.substring(1) : tagName

const printVersions = (list, message) => {
    log(message)
    list.forEach(item => {
        log(`  - ${item}`)
    })
}

const getVersionsFromTags = () => {
    const options = {
        url: 'https://api.github.com/repos/yarnpkg/yarn/tags',
        headers: {
            'User-Agent': 'YVM',
        },
    }

    return new Promise((resolve, reject) => {
        request.get(options, (error, response, body) => {
            if (error || response.statusCode !== 200) {
                reject(error)
            } else {
                const tags = JSON.parse(body)
                const tagNames = tags.map(tag => tag.name)
                const versions = tagNames
                    .map(stripVersionPrefix)
                    .filter(version => version[0] > 0)
                resolve(versions)
            }
        })
    })
}

module.exports = {
    getExtractionPath,
    getVersionsFromTags,
    printVersions,
    stripVersionPrefix,
    versionRootPath,
    yvmPath,
}
