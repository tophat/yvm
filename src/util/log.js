const chalk = require('chalk')

const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

const log = (...args) => log.default(...args)

log.default = function defaultLog(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

log.info = function infoLog(...args) {
    if (process.env.YVM_VERBOSE) {
        log(...args)
    }
}

log.error = function errorLog(...args) {
    log(chalk.red(...args))
}

log.success = function successLog(...args) {
    log(chalk.green(...args))
}

log.notice = function noticeLog(...args) {
    log(chalk.yellow(...args))
}

module.exports = log
