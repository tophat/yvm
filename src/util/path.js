const os = require('os')
const path = require('path')

const yvmPath = process.env.YVM_DIR || path.resolve(os.homedir(), '.yvm')
module.exports = { yvmPath }
