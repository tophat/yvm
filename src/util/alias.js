import fs from 'fs'
import { execSync } from 'child_process'
import path from 'path'

import { memoize } from 'lodash'
import chalk from 'chalk'

import log from 'util/log'
import { yvmPath as defaultYvmPath, getNonYvmYarnPathEntries } from 'util/path'
import { getRequest, getVersionsFromTags } from 'util/utils'
import { YARN_STABLE_VERSION_URL } from 'util/constants'

const filterAliasesByName = (pattern, aliases) => {
    return Object.keys(aliases)
        .filter(name => new RegExp(pattern).test(name))
        .map(name => ({ name, value: aliases[name] }))
}

const getAllDependants = ({ name, aliases }) => {
    const visited = []
    const dependants = [name]
    const aliasNames = Object.keys(aliases)
    while (dependants.length) {
        const aliasName = dependants.shift()
        if (visited.includes(aliasName)) {
            continue
        }
        visited.push(aliasName)
        dependants.push(...aliasNames.filter(d => aliases[d] === aliasName))
    }
    return visited
}

export const STORAGE_FILE = '.aliases'

export const DEFAULT = 'default'
export const LATEST = 'latest'
export const LTS = 'lts' // TODO: resolve long term support version
export const STABLE = 'stable'
export const SYSTEM = 'system'
export const UNRESOLVED = undefined
export const NOT_AVAILABLE = 'N/A'
export const DEFAULT_NAMES = [LATEST, STABLE, SYSTEM]
export const RESERVED_NAMES = [...DEFAULT_NAMES, LTS, NOT_AVAILABLE]

export const resolveLatest = memoize(
    async () => (await getVersionsFromTags())[0],
)

export const resolveStable = memoize(async () => {
    const version = await getRequest(YARN_STABLE_VERSION_URL)
    return (version && version.trim()) || UNRESOLVED
})

export const resolveSystem = memoize(async ({ shell } = {}) => {
    for (const pathToExec of getNonYvmYarnPathEntries(shell)) {
        try {
            const execCommand = `${pathToExec} --version 2> /dev/null`
            return String(execSync(execCommand)).trim()
        } catch (error) {
            log.info(error.message)
        }
    }
    return NOT_AVAILABLE
})

export const isReserved = name => RESERVED_NAMES.includes(name)

export const resolveReserved = memoize(async (name, args = {}) => {
    const resolver =
        {
            [LATEST]: resolveLatest,
            [STABLE]: resolveStable,
            [SYSTEM]: resolveSystem,
        }[name] || (async () => NOT_AVAILABLE)
    return resolver(args)
})

export const getDefaultAliases = () => {
    return DEFAULT_NAMES.reduce(
        (aliases, name) => Object.assign(aliases, { [name]: UNRESOLVED }),
        {},
    )
}

export const getUserAliases = memoize(async (yvmPath = defaultYvmPath) => {
    const aliases = { [DEFAULT]: STABLE }
    const aliasStoragePath = path.join(yvmPath, STORAGE_FILE)
    try {
        const aliasesJSON = fs.readFileSync(aliasStoragePath, 'utf8')
        Object.assign(aliases, JSON.parse(aliasesJSON))
    } catch (e) {
        log.info(e.message)
    }
    return aliases
})

export const getMatchingAliases = async (pattern, yvmPath = defaultYvmPath) => {
    return filterAliasesByName(pattern, await getUserAliases(yvmPath))
}

export const setAlias = async ({ name, version, yvmPath = defaultYvmPath }) => {
    const aliasStoragePath = path.join(yvmPath, STORAGE_FILE)
    try {
        if (isReserved(name)) {
            throw new Error(`'${name}' is reserved. Please choose another.`)
        }
        const aliases = await getUserAliases(yvmPath)
        aliases[name] = version
        fs.writeFileSync(aliasStoragePath, JSON.stringify(aliases))
        return true
    } catch (e) {
        log(`Unable to set alias. ${e.message}`)
        log.info(e.stack)
        return false
    }
}

export const unsetAlias = async ({
    name,
    force = false,
    recursive = false,
    yvmPath = defaultYvmPath,
}) => {
    const aliasStoragePath = path.join(yvmPath, STORAGE_FILE)
    try {
        if (isReserved(name)) {
            throw new Error(`'${name}' is reserved by yvm.`)
        }
        let deleted = false
        const aliases = await getUserAliases(yvmPath)
        const aliasNameList = Object.keys(aliases)
        const dependants = aliasNameList.filter(d => aliases[d] === name)
        if (dependants.length && !force && !recursive) {
            log.notice(`The following aliases will be orphaned: ${dependants}
Rerun with '--force' to remove just this alias
Or with '--recursive' to remove all dependants`)
        } else {
            const namesToRemove = recursive
                ? getAllDependants({ name, aliases })
                : [name]
            const newAliases = {}
            for (const aliasName of aliasNameList) {
                if (namesToRemove.includes(aliasName)) continue
                newAliases[aliasName] = aliases[aliasName]
            }
            deleted = aliases[name] !== newAliases[name]
            log.info(`The following aliases were removed: ${namesToRemove}`)
            fs.writeFileSync(aliasStoragePath, JSON.stringify(newAliases))
        }
        if (dependants.length && deleted && !recursive) {
            log.notice(`The following aliases are now orphaned: ${dependants}`)
        }
        return deleted
    } catch (e) {
        log(`Unable to unset ${e.message}`)
        log.info(e.stack)
        return false
    }
}

export const resolveAlias = memoize(
    async ({
        versionString,
        currentAliases,
        visitedAliases = [],
        yvmPath = defaultYvmPath,
    }) => {
        const chain = visitedAliases.join(', ')
        const resolveErrorMessage = msg => `${msg}: '${versionString}' ${chain}`
        if (visitedAliases.includes(versionString)) {
            throw new Error(resolveErrorMessage`Cyclic chain detected`)
        }
        const aliases = currentAliases || (await getUserAliases(yvmPath))
        if (versionString in aliases && !isReserved(versionString)) {
            return await resolveAlias({
                versionString: aliases[versionString],
                currentAliases: aliases,
                visitedAliases: [...visitedAliases, versionString],
                yvmPath,
            })
        }
        return {
            visited: visitedAliases,
            version: versionString,
        }
    },
)

export const resolveAliases = memoize(async (yvmPath = defaultYvmPath) => {
    const currentAliases = {
        ...(await getUserAliases(yvmPath)),
        ...getDefaultAliases(),
    }
    const resolvedAliases = {}
    for (const versionString of Object.keys(currentAliases)) {
        if (versionString in resolvedAliases) {
            continue
        }
        const { version, visited } = await resolveAlias({
            versionString,
            currentAliases,
            yvmPath,
        }).catch(() => ({ version: undefined, visited: [] }))
        for (const alias of [versionString, ...visited]) {
            resolvedAliases[alias] = {
                value: currentAliases[alias],
                version,
            }
        }
    }
    return resolvedAliases
})

export const resolveMatchingAliases = memoize(
    async (pattern = '', yvmPath = defaultYvmPath) => {
        return filterAliasesByName(pattern, await resolveAliases(yvmPath))
    },
)

export const getFormatter = (
    allVersions,
    installedVersions,
    currentVersion,
) => (name, version, target) => {
    const defaultStyler = chalk.grey
    let termStyler = defaultStyler
    const isAvailable = allVersions.includes(target)
    if (target === currentVersion) {
        termStyler = chalk.green
    } else if (installedVersions.includes(target)) {
        termStyler = chalk.yellow
    } else if (isAvailable) {
        termStyler = chalk.white
    } else {
        termStyler = chalk.red
    }
    if (isReserved(name)) {
        termStyler = termStyler.bold
    }
    const parts = [termStyler(name), ' â†’ ']
    const targetVersion = termStyler(isAvailable ? target : NOT_AVAILABLE)
    if (version) {
        parts.push(chalk.white(version), ' (', targetVersion, ')')
    } else {
        parts.push(targetVersion)
    }

    return defaultStyler(parts.join(''))
}
