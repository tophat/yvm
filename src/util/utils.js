import path from 'path'
import request from 'request'
import memoize from 'lodash/memoize'

import {
    USER_AGENT,
    YARN_DOWNLOAD_URL,
    YARN_RELEASES_API_URL,
} from 'util/constants'
import log from 'util/log'

export const versionRootPath = rootPath => path.resolve(rootPath, 'versions')

export const getExtractionPath = (version, rootPath) =>
    path.resolve(rootPath, 'versions', `v${version}`)

export const stripVersionPrefix = tagName =>
    tagName[0] === 'v' ? tagName.substring(1) : tagName

export const getRequest = memoize(async url => {
    const options = {
        url,
        gzip: true,
        headers: {
            'User-Agent': USER_AGENT,
        },
    }
    return new Promise((resolve, reject) => {
        request.get(options, (error, response, body) => {
            const { statusCode, body: responseBody } = response || {}
            if (error || statusCode !== 200) {
                if (responseBody) {
                    if (error) {
                        log(error)
                    }
                    return reject(responseBody)
                }
                return reject(error)
            }
            return resolve(body)
        })
    })
})

export const getVersionDownloadUrl = version =>
    `${YARN_DOWNLOAD_URL}/${version}/yarn-v${version}.tar.gz`

export const getReleasesFromTags = memoize(async () => {
    return getRequest(YARN_RELEASES_API_URL).then(body => {
        return JSON.parse(body).reduce((accumulator, tag) => {
            const version = stripVersionPrefix(tag.name)
            const [major] = version.split('.')
            return Number(major) > 0
                ? Object.assign(accumulator, { [version]: tag })
                : accumulator
        }, {})
    })
})

export const getVersionsFromTags = memoize(async () => {
    try {
        return Object.keys(await getReleasesFromTags())
    } catch (e) {
        log.error(
            'Unable to retrieve remote versions. Please check your network connection',
        )
        return []
    }
})
