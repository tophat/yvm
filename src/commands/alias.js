import log from '../util/log'
import { getFormatter, resolveMatchingAliases, setAlias } from '../util/alias'
import {
    getVersionInUse,
    getYarnVersions,
    resolveVersion,
} from '../util/version'
import { getVersionsFromTags } from '../util/utils'

export const alias = async (nameOrPattern, maybeVersion) => {
    try {
        const format = getFormatter(
            await getVersionsFromTags(),
            getYarnVersions(),
            await getVersionInUse(),
        )

        const safeResolveVersion = async versionString =>
            resolveVersion({ versionString }).catch(e => log.info(e.message))

        if (typeof maybeVersion === 'string') {
            const name = nameOrPattern
            const version = maybeVersion
            let targetVersion
            if (await setAlias({ name, version })) {
                targetVersion = await safeResolveVersion(version)
            }
            const message = format(name, version, targetVersion)
            if (targetVersion) {
                log(message)
                return 0
            } else {
                log.error(message)
                return 1
            }
        } else {
            const pattern = nameOrPattern || ''
            const matchingAliases = await resolveMatchingAliases(pattern)
            for (const {
                name,
                value: { value, version: versionString },
            } of matchingAliases) {
                const version = await safeResolveVersion(versionString)
                log(format(name, value, version))
            }
            const noMatchFound =
                pattern.length > 0 && matchingAliases.length < 1
            return noMatchFound ? 1 : 0
        }
    } catch (e) {
        log.error(e.message)
        return 2
    }
}
