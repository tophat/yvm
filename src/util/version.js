import fs from 'fs'
import { execSync } from 'child_process'
import semver from 'semver'
import cosmiconfig from 'cosmiconfig'
import memoize from 'lodash/memoize'

import log from 'util/log'
import { yvmPath as defaultYvmPath } from 'util/path'
import {
    getVersionsFromTags,
    stripVersionPrefix,
    versionRootPath,
} from 'util/utils'
import {
    DEFAULT,
    isReserved,
    resolveAlias,
    resolveReserved,
    setAlias,
} from 'util/alias'

export const DEFAULT_VERSION_TEXT = 'Global Default'
export const VERSION_IN_USE_SYMBOL = '\u2713'
export const VERSION_INSTALLED_SYMBOL = '\u2192'

export const isValidVersionString = version =>
    semver.valid(version.trim()) !== null
export const isValidVersionRange = versionRange => {
    return semver.validRange(versionRange.trim()) !== null
}
export const getValidVersionString = version => semver.clean(version)

/**
 * Get the most recent yarn install version in the specified range
 * @param {String} versionRange the bounding yarn version range
 * @throws if no install version can be found for range
 */
export const getVersionFromRange = memoize(async versionRange => {
    const availableVersion =
        semver.maxSatisfying(getYarnVersions(), versionRange) ||
        semver.maxSatisfying(await getVersionsFromTags(), versionRange)
    if (!availableVersion) {
        throw new Error(
            `Given version can not be satisfied by yarn: ${versionRange}
See list of available yarn versions using: 'yvm list-remote'`,
        )
    }
    return availableVersion
})

export const resolveVersion = memoize(
    async ({ versionString, yvmPath = defaultYvmPath }) => {
        const { version } = await resolveAlias({ versionString, yvmPath })
        if (version) {
            const validVersion = isValidVersionString(version)
            const validVersionRange = isValidVersionRange(version)
            if (validVersion || validVersionRange) {
                return await getVersionFromRange(version)
            }
            if (isReserved(version)) {
                return await resolveReserved(version, { yvmPath })
            }
        }
        throw new Error(`Unable to resolve: ${versionString}`)
    },
)

export const getDefaultVersion = async (yvmPath = defaultYvmPath) => {
    try {
        return await resolveVersion({
            versionString: DEFAULT,
            yvmPath,
        })
    } catch (e) {
        log.info(e)
    }
}

export const setDefaultVersion = async ({
    version,
    yvmPath = defaultYvmPath,
}) => {
    try {
        await setAlias({ name: DEFAULT, version, yvmPath })
        return true
    } catch (e) {
        log('Unable to set default version')
        log.info(e)
        return false
    }
}

export const getRcFileVersion = () => {
    const moduleName = 'yvm'
    const explorer = cosmiconfig(moduleName, {
        packageProp: 'engines.yarn',
        searchPlaces: [
            'package.json',
            `.${moduleName}rc`,
            `.${moduleName}rc.json`,
            `.${moduleName}rc.yaml`,
            `.${moduleName}rc.yml`,
            `.${moduleName}rc.js`,
            `${moduleName}.config.js`,
            `.yarnversion`,
        ],
    })
    const result = explorer.searchSync()
    if (!result || result.isEmpty || !result.config) {
        return null
    }
    log.info(`Found config ${result.filepath}`)
    return String(result.config)
}

export const getVersionInUse = memoize(async () => {
    try {
        return String(execSync('yarn --version')).trim()
    } catch (error) {
        log.info(error)
        return ''
    }
})

export const getYarnVersions = memoize((yvmPath = defaultYvmPath) => {
    const versionsPath = versionRootPath(yvmPath)
    if (fs.existsSync(versionsPath)) {
        const files = fs.readdirSync(versionsPath)
        return files
            .filter(name => name.startsWith('v') && isValidVersionString(name))
            .map(stripVersionPrefix)
    }
    return []
})

export const getSplitVersionAndArgs = async (maybeVersionArg, ...rest) => {
    let versionToUse
    try {
        if (maybeVersionArg) {
            log.info(`Attempting to resolve ${maybeVersionArg}`)
            const parsedVersionString = await resolveVersion({
                versionString: maybeVersionArg,
            }).catch(e => {
                log.info(e.stack)
            })
            if (parsedVersionString) {
                log.info(`Using provided version: ${parsedVersionString}`)
                return [parsedVersionString, rest]
            }
            rest.unshift(maybeVersionArg)
        }

        const rcVersion = getRcFileVersion()
        if (rcVersion) {
            log.info(`Resolving version '${rcVersion}' found in config`)
            versionToUse = await resolveVersion({
                versionString: rcVersion,
            })
        } else {
            log.info(`Attempting to use default version.`)
            versionToUse = await getDefaultVersion()
        }
    } catch (e) {
        log.error(e.message)
        log.info(e.stack)
    }

    if (!versionToUse) {
        throw new Error(
            `No yarn version supplied!
Try:
    Setting a global default version (yvm set-default 1.9.2)
    Adding .yvmrc file in this directory or a parent directory
    Specify your version in the command.`,
        )
    }
    log.info(`Using yarn version: ${versionToUse}`)
    return [versionToUse, rest]
}

export const printVersions = async ({
    list,
    message,
    versionInUse = '',
    defaultVersion,
    localVersions = [],
}) => {
    log(message)

    versionInUse = versionInUse.trim()
    defaultVersion = defaultVersion || (await getDefaultVersion(defaultYvmPath))

    const versionsMap = {}

    list.forEach(versionPadded => {
        const version = versionPadded.trim()
        const isCurrent = version === versionInUse
        const isDefault = version === defaultVersion
        const isInstalled = localVersions.includes(version)

        let toLog = ' '
        if (isCurrent) toLog += VERSION_IN_USE_SYMBOL
        else if (isInstalled) toLog += VERSION_INSTALLED_SYMBOL
        else toLog += '-'
        toLog += ` ${versionPadded}`

        if (isDefault) toLog += ` (${DEFAULT_VERSION_TEXT})`

        if (isCurrent) log.success(toLog)
        else if (isInstalled) log.notice(toLog)
        else log(toLog)

        versionsMap[version] = toLog
    })
    return versionsMap
}
