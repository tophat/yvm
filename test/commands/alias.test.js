import fs from 'fs-extra'
import path from 'path'

import { alias } from 'commands/alias'
import { STORAGE_FILE, getUserAliases } from 'util/alias'
import log from 'util/log'
import { yvmPath as rootPath } from 'util/path'
import * as utils from 'util/utils'
import * as version from 'util/version'

jest.mock('util/path', () => ({
    yvmPath: '/tmp/cmd/alias/yvm',
    getNonYvmYarnPathEntries: () => [],
}))

describe('alias', () => {
    const { resolveVersion } = version
    const currentYarnVersion = '1.15.2'
    const pinnedStableVersion = currentYarnVersion
    const installedYarnVersions = [currentYarnVersion, '1.13.0', '1.7.0']
    const allYarnVersions = [...installedYarnVersions, '1.6.0', '1.3.0']
    const mockAliases = {
        abc: 'b',
        bcd: 'z',
        xyz: '1.13',
    }

    const writeAliases = () => {
        const aliasFilePath = path.join(rootPath, STORAGE_FILE)
        fs.outputFileSync(aliasFilePath, JSON.stringify(mockAliases))
    }

    beforeAll(() => {
        jest.spyOn(version, 'resolveVersion').mockImplementation(
            async (...args) => {
                if (args[0].versionString === 'stable')
                    return pinnedStableVersion
                return resolveVersion(...args)
            },
        )
        jest.spyOn(version, 'getVersionInUse').mockResolvedValue(
            currentYarnVersion,
        )
        jest.spyOn(version, 'getYarnVersions').mockReturnValue(
            installedYarnVersions,
        )
        jest.spyOn(utils, 'getVersionsFromTags').mockResolvedValue(
            allYarnVersions,
        )
        jest.spyOn(log, 'default')
        jest.spyOn(log, 'error')
    })

    beforeEach(() => {
        writeAliases()
        getUserAliases.cache.clear()
        jest.clearAllMocks()
    })

    afterAll(jest.restoreAllMocks)

    describe('setAlias', () => {
        it('sets alias correctly', async () => {
            expect(await alias('def', '1.15')).toBe(0)
            expect(log.default.mock.calls).toMatchSnapshot()
            expect(await getUserAliases()).toMatchObject({
                ...mockAliases,
                def: '1.15',
            })
        })

        it('handles invalid version', async () => {
            expect(await alias('def', 'invalid')).toBe(1)
            expect(log.default.mock.calls).toMatchSnapshot()
            expect(await getUserAliases()).toMatchObject({
                ...mockAliases,
                def: 'invalid',
            })
        })

        it('handles thrown failure', async () => {
            const mockError = new Error('some error')
            utils.getVersionsFromTags.mockRejectedValueOnce(mockError)
            expect(await alias('def', '1.15')).toBe(2)
            expect(log.error).toHaveBeenCalledWith(mockError.message)
        })
    })

    describe('getAlias', () => {
        it('matches all aliases with empty pattern', async () => {
            expect(await alias()).toBe(0)
            expect(log.default.mock.calls).toMatchSnapshot()
        })

        it('matches only supplied pattern', async () => {
            expect(await alias('bc')).toBe(0)
            expect(log.default.mock.calls).toMatchSnapshot()
        })

        it('handles no matches for supplied pattern', async () => {
            expect(await alias('abracadabra')).toBe(1)
            expect(log.default.mock.calls).toMatchSnapshot()
        })

        it('handles thrown failure', async () => {
            const mockError = new Error('some error')
            utils.getVersionsFromTags.mockRejectedValueOnce(mockError)
            expect(await alias()).toBe(2)
            expect(log.error).toHaveBeenCalledWith(mockError.message)
        })
    })
})
