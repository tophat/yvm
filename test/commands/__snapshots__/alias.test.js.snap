// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alias command getAlias handles no matches for supplied pattern 1`] = `Array []`;

exports[`alias command getAlias matches all aliases with empty pattern 1`] = `
Array [
  Array [
    "[90m[32mdefault[39m[90m â†’ [37mstable[39m[90m ([32m1.15.2[39m[90m)[39m",
  ],
  Array [
    "[90m[31mabc[39m[90m â†’ [37mb[39m[90m ([31mN/A[39m[90m)[39m",
  ],
  Array [
    "[90m[31mbcd[39m[90m â†’ [37mz[39m[90m ([31mN/A[39m[90m)[39m",
  ],
  Array [
    "[90m[33mxyz[39m[90m â†’ [37m1.13[39m[90m ([33m1.13.0[39m[90m)[39m",
  ],
  Array [
    "[90m[32m[1mlatest[22m[39m[90m â†’ [32m[1m1.15.2[22m[39m[90m[39m",
  ],
  Array [
    "[90m[32m[1mstable[22m[39m[90m â†’ [32m[1m1.15.2[22m[39m[90m[39m",
  ],
  Array [
    "[90m[31m[1msystem[22m[39m[90m â†’ [31m[1mN/A[22m[39m[90m[39m",
  ],
]
`;

exports[`alias command getAlias matches only supplied pattern 1`] = `
Array [
  Array [
    "[90m[31mabc[39m[90m â†’ [37mb[39m[90m ([31mN/A[39m[90m)[39m",
  ],
  Array [
    "[90m[31mbcd[39m[90m â†’ [37mz[39m[90m ([31mN/A[39m[90m)[39m",
  ],
]
`;

exports[`alias command setAlias handles invalid version 1`] = `
Array [
  Array [
    "[31m[90m[31mdef[39m[31m[90m â†’ [37minvalid[39m[31m[90m ([31mN/A[39m[31m[90m)[39m[31m[39m",
  ],
]
`;

exports[`alias command setAlias sets alias correctly 1`] = `
Array [
  Array [
    "[90m[32mdef[39m[90m â†’ [37m1.15[39m[90m ([32m1.15.2[39m[90m)[39m",
  ],
]
`;
